tuf/scheme/ecdsa_sha256.go

```diff
@@ -37,15 +50,124 @@ func NewECDSA() *ECDSA {
 	}
 	return &ECDSA{
 		PrivateKey: privateKey,
-		PublicKey:  privateKey.Public(),
-		KeyType:    ECDSA256,
+		KeyType:    types.ECDSA256,
 	}
 }
 
-func (ecdsa *ECDSA) Store(filename string) error {
-	keyJson, err := json.Marshal(ecdsa)
+func (ecdsaKey *ECDSA) encode() (string, string, error) {
+	x509Encoded, err := x509.MarshalECPrivateKey(ecdsaKey.PrivateKey)
+	if err != nil {
+		return "", "", err
+	}
+	pemEncoded := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: x509Encoded})
+
+	x509EncodedPub, err := x509.MarshalPKIXPublicKey(&ecdsaKey.PrivateKey.PublicKey)
+	if err != nil {
+		return "", "", err
+	}
+	pemEncodedPub := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: x509EncodedPub})
+	return string(pemEncoded), string(pemEncodedPub), nil
+}
+
+func (ecdsaKey *ECDSA) decode(pemEncoded string) error {
+	block, _ := pem.Decode([]byte(pemEncoded))
+	privateKey, err := x509.ParseECPrivateKey(block.Bytes)
 	if err != nil {
 		return err
 	}
-	return ioutil.WriteFile(filename, keyJson, 0644)
+	ecdsaKey.PrivateKey = privateKey
+	ecdsaKey.KeyType = types.ECDSA256
+	return nil
+}
+
+func (ecdsaKey *ECDSA) Store(filename string) error {
+	privateKey, publicKey, err := ecdsaKey.encode()
+	schemeKey := SchemeKey{
+		PrivateKey: privateKey,
+		PublicKey:  publicKey,
+		KeyType:    ecdsaKey.KeyType,
+	}
+	jsonBytes, err := json.Marshal(schemeKey)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, jsonBytes, 0644)
+}
+
+func (ecdsaKey *ECDSA) Sign(singedMetadata interface{}) (string, error) {
+	// Convert singedMetadata to bytes.
+	var buf bytes.Buffer
+	err := gob.NewEncoder(&buf).Encode(singedMetadata)
+	if err != nil {
+		return "", err
+	}
+	fmt.Println("while encoding gob encoder bytes", buf.Bytes())
+	// Calculate hash of string using SHA256 algo because only first 32 bytes are verified.
+	sha256Sum := sha256.Sum256(buf.Bytes())
+	r, s, err := ecdsa.Sign(rand.Reader, ecdsaKey.PrivateKey, sha256Sum[0:])
+	if err != nil {
+		return "", err
+	}
+	curveBits := ecdsaKey.Curve.Params().BitSize
+
+	if CURVESIZE != curveBits {
+		return "", ErrInvalidKey
+	}
+	keyBytes := KEYSIZE
+	if curveBits%8 > 0 {
+		keyBytes++
```

Comment:
I'm not sure I understand the need for this part.

Reply:
Basically this all the way up through the hex.EncodeToString. Are you following any examples here?

Reply:
I've seen some examples that just append the bytes together from each and skip the 32 bit integer. Would that work?

Reply:
done